Core/src/main/java/org/tribuo/protos/ProtoUtil.java

```diff
@@ -0,0 +1,474 @@
+            ProtoSerializableClass annotation = protoSerializable.getClass().getAnnotation(ProtoSerializableClass.class);
+            if (annotation == null) {
+                throw new IllegalArgumentException("instance of ProtoSerializable must be annotated with @ProtoSerializableClass to be serialized with ProtoUtil.serialize()");
+            }
+
+            Class<SERIALIZED_CLASS> serializedClass = getSerializedClass(protoSerializable);
+            SERIALIZED_CLASS.Builder serializedClassBuilder = (SERIALIZED_CLASS.Builder) serializedClass.getMethod("newBuilder").invoke(null);
+            Class<SERIALIZED_CLASS.Builder> serializedClassBuilderClass = (Class<SERIALIZED_CLASS.Builder>) serializedClassBuilder.getClass();
+            serializedClassBuilderClass.getMethod("setVersion", Integer.TYPE).invoke(serializedClassBuilder, annotation.version());
+            serializedClassBuilderClass.getMethod("setClassName", String.class).invoke(serializedClassBuilder, protoSerializable.getClass().getName());
+
+            Class<SERIALIZED_DATA> serializedDataClass = (Class<SERIALIZED_DATA>) annotation.serializedDataClass();
+            //the default value for ProtoSerializableClass.serializedDataClass is GeneratedMessageV3 which is how you can signal that
+            //there is no serialized data to be serialized.
+            if (serializedDataClass != Message.class) {
+                SERIALIZED_DATA.Builder serializedDataBuilder = (SERIALIZED_DATA.Builder) serializedDataClass.getMethod("newBuilder").invoke(null);
+                Class<SERIALIZED_DATA.Builder> serializedDataBuilderClass = (Class<SERIALIZED_DATA.Builder>) serializedDataBuilder.getClass();
+
+                for (Field field : getFields(protoSerializable.getClass())) {
+                    field.setAccessible(true);
+                    Object obj = field.get(protoSerializable);
+
+                    ProtoSerializableField protoSerializableField = field.getAnnotation(ProtoSerializableField.class);
+                    if (protoSerializableField != null) {
+                        String fieldName = protoSerializableField.name();
+                        if (fieldName.equals(ProtoSerializableField.DEFAULT_FIELD_NAME)) {
+                            fieldName = field.getName();
+                        }
+
+                        Class<?> fieldClazz = field.getType();
+                        String nameStub = (fieldClazz.isArray() || Collection.class.isAssignableFrom(fieldClazz)) ? "addAll" : "set";
+                        Method setter = findMethod(serializedDataBuilderClass, nameStub, fieldName, 1);
+                        obj = convert(obj);
+                        setter.setAccessible(true);
+                        setter.invoke(serializedDataBuilder, obj);
+                        setter.setAccessible(false);
+                    }
+
+                    ProtoSerializableKeysValuesField pskvf = field.getAnnotation(ProtoSerializableKeysValuesField.class);
```

Comment:
I don't think a field should ever be more than one of `ProtoSerializableField`, `ProtoSerializableKeysValuesField`, or `ProtoSerializableMapValuesField`, and this seems like the place to enforce that.

Reply:
Will do.

Reply:
It looks like this is actually checked it `getFields` further down

Reply:
Looks like this is accounted for in `getFields`